smart-contract-verifier/src/compiler/fetcher.rs

```diff
@@ -106,14 +110,39 @@ mod tests {
     #[tokio::test]
     async fn write_text_executable() {
         let tmp_dir = tempfile::tempdir().unwrap();
+
         let data = "this is a compiler binary";
         let bytes = Bytes::from_static(data.as_bytes());
         let sha = Sha256::digest(data.as_bytes());
+
         let version = Version::from_str("v0.4.10+commit.f0d539ae").unwrap();
         let file = write_executable(bytes, H256::from_slice(&sha), tmp_dir.path(), &version)
             .await
             .unwrap();
+
         let content = tokio::fs::read_to_string(file).await.unwrap();
         assert_eq!(data, content);
     }
+
+    #[tokio::test]
+    async fn wrong_file_checksum() {
+        let tmp_dir = tempfile::tempdir().unwrap();
+
+        let data = "this is a compiler binary";
+        let bytes = Bytes::from_static(data.as_bytes());
+        let sha = H256::default();
+        let version = Version::from_str("v0.4.10+commit.f0d539ae").unwrap();
+
+        let err = write_executable(bytes, sha, tmp_dir.path(), &version)
+            .await
+            .expect_err("expected to fail with wrong checksum");
+        assert!(matches!(err, FetchError::HashMismatch(_)));
+
+        let dir = tmp_dir.path().join(version.to_string());
+        let file = dir.join("solc");
+        let mut tmp_file = file.clone();
+        tmp_file.set_extension("tmp");
+        assert!(!file.exists());
+        assert!(tmp_file.exists());
```

Comment:
Shouldn't temporary file be deleted if checksum was invalid?

Reply:
we can make this, but there still will be cases, where tmp file could exists, for example if server will shutdown while saving the file.

we can try to delete file in some places, but I don't see the reason to do so, because we can't delete it in every case.  Specially seeing that checksum mismatch is something rare

Reply:
Okay